docs/src/adr/014adr-precise-records.md

```diff
@@ -0,0 +1,670 @@
+  log
+```
+
+In this case, we would not be able to easily use `FilterByTag`. Of course, we
+could wrap a variant into a singleton set and then apply `FilterByTag` to it
+and `CHOOSE` on top of it. However, this looks clunky and does not guarantee
+type safety. A much simpler solution is to introduce another special operator:
+
+```tla
+MatchTag(variant, tag, ThenOper(_), ElseOper(_)) ==
+  \* fallback untyped implementation
+  IF variant.tag = tag
+  THEN ThenOper(variant)
+  ELSE ElseOper(variant)
+```
+
+The idea of `MatchTag` is that it passes the extracted record to `ThenOper`,
+when its tag value matches `tag`; otherwise, it passes the reduced variant to
+`ElseOper`. This is precisely captured by the inference rule:
+
+```
+variant: Variant(RowCons("<TAG>", r, t))
+ThenOper: r => z
+ElseOper: Variant(t) => z
+------------------------------------------------ [variant_match]
+MatchTag(variant, "<TAG>", ThenOper, ElseOper):
+  (Variant(RowCons("<TAG>", r, t)),
+   Str,
+   r => z,
+   Variant(t) => z
+  ) => z
+```
+
+The above rule looks menacing. Here is an example of matching a record
+in the above example with the variable `log`:
+
+```tla
+IsDefined(eventAOrB) ==
+  LET ElseB(onlyB) ==
+    MatchTag(onlyB, "EventB", LAMBDA b: b.src /= "", LAMBDA elseValue: FALSE)
+  IN  
+  MatchTag(event, "EventA", LAMBDA a: a.val /= -1, ElseB)
+```
+
+The operator `ElseB` looks redundant, as we know that `onlyB` is a singleton
+variant. To this end, we introduce the operator `MatchOnly`:
+
+```tla
+MatchOnly(variant, ThenOper(_)) ==
+  \* fallback untyped implementation
+  ThenOper(variant)
+```
+
+For completeness, we give the inference rule for `MatchOnly`:
+
+```
+variant: Variant(RowCons("<TAG>", r, RowNil))
+ThenOper: r => z
+------------------------------------------------ [variant_match_only]
+MatchOnly(variant, ThenOper):
+  (Variant(RowCons("<TAG>", r, RowNil)),
+   r => z
+  ) => z
+```
+
+It looks like the solution with `MatchTag` and `MatchOnly` introduce a lot of
+boilerplate. However, this is probably the best solution that we can have,
+unless we can extend the grammar of TLA+.
+
+### 5.3. Changes in the model checker
+
+Having precise types for variants, we have two options:
+
+ 1. Keep the current encoding, that is, a variant is encoded as a super-record
+ that contains all possible fields of the member records. The type checker
+ will guarantee that we do not access the fields of the super-record that are
+ not present in the actual record type.
+
+ 1. Implement the
+ [suggestion by Shon Feder](https://github.com/informalsystems/apalache/discussions/789#discussioncomment-1592118).
+
+Although Option 2 looks nicer, we prefer keeping Option 1. The reason is that
+the current implementation introduces the minimal number of constraints
+by mashing all possible fields into a super-record. The alternative solution
+(option 2) would introduce additional constraints, when the spec requires us
+to extract an element from a set.
```

Comment:
Notably, with the precise record types from Snowcat, it's trivial to trim records in post-processing.

Reply:
I am not sure, what you mean. The hardest case is `\E x \in S: ...`, where `S` is a set of variants. Since it is impossible to statically compute the exact type of `x`, you would have to deal with tuples in their generality.

Reply:
What I mean is, if you have the variant operators guiding type-checking, you can use those types to reconstruct the smt model values of record type.

Reply:
We are constructing a system of equalities that are solved by unification. Are you talking about your branch on type inference with SMT?

Reply:
No, I'm talking about cases where the current type checker infers a record with spurious fields, and then cannot catch illegal field access. In the post-ADR world, you could use variant information to avoid this.

Reply:
Yes, exactly. So the point is that the arena encoding is still using spurious fields, but by knowing types we can guarantee that spurious fields are never accessed.